File: android/src/main/kotlin/com/instrumentisto/medea_flutter_webrtc/controller/PeerConnectionController.kt
package com.instrumentisto.medea_flutter_webrtc.controller

import com.instrumentisto.medea_flutter_webrtc.model.*
import com.instrumentisto.medea_flutter_webrtc.proxy.MediaStreamTrackProxy
import com.instrumentisto.medea_flutter_webrtc.proxy.PeerConnectionProxy
import com.instrumentisto.medea_flutter_webrtc.proxy.RtpTransceiverProxy
import com.instrumentisto.medea_flutter_webrtc.utils.AnyThreadSink
import com.instrumentisto.medea_flutter_webrtc.utils.resultUnhandledException
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch

/**
 * Controller of [PeerConnectionProxy] functional.
 *
 * @property messenger Messenger used for creating new [MethodChannel]s.
 * @property peer Underlying [MediaStreamTrackProxy] to perform [MethodCall]s on.
 */
class PeerConnectionController(
    private val messenger: BinaryMessenger,
    private val peer: PeerConnectionProxy
) : EventChannel.StreamHandler, Controller {
  /** [CoroutineScope] for this [PeerConnectionController] */
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

  /** Unique ID of the [MethodChannel] of this controller. */
  private val channelId = nextChannelId()

  /** Channel listened for the [MethodCall]s. */
  private val chan: MethodChannel =
      MethodChannel(messenger, ChannelNameGenerator.name("PeerConnection", channelId))

  /** Event channel into which all [PeerConnectionProxy] events are sent. */
  private val eventChannel: EventChannel =
      EventChannel(messenger, ChannelNameGenerator.name("PeerConnectionEvent", channelId))

  /** Event sink into which all [PeerConnectionProxy] events are sent. */
  private var eventSink: AnyThreadSink? = null

  /** [PeerConnectionProxy] events observer which sends all events to the [eventSink]. */
  private val eventObserver =
      object : PeerConnectionProxy.Companion.EventObserver {
        override fun onTrack(track: MediaStreamTrackProxy, transceiver: RtpTransceiverProxy) {
          eventSink?.success(
              mapOf(
                  "event" to "onTrack",
                  "track" to MediaStreamTrackController(messenger, track).asFlutterResult(),
                  "transceiver" to
                      RtpTransceiverController(messenger, transceiver).asFlutterResult()))
        }

        override fun onIceConnectionStateChange(iceConnectionState: IceConnectionState) {
          eventSink?.success(
              mapOf("event" to "onIceConnectionStateChange", "state" to iceConnectionState.value))
        }

        override fun onSignalingStateChange(signalingState: SignalingState) {
          eventSink?.success(
              mapOf("event" to "onSignalingStateChange", "state" to signalingState.value))
        }

        override fun onConnectionStateChange(peerConnectionState: PeerConnectionState) {
          eventSink?.success(
              mapOf("event" to "onConnectionStateChange", "state" to peerConnectionState.value))
        }

        override fun onIceGatheringStateChange(iceGatheringState: IceGatheringState) {
          eventSink?.success(
              mapOf("event" to "onIceGatheringStateChange", "state" to iceGatheringState.value))
        }

        override fun onIceCandidate(candidate: IceCandidate) {
          eventSink?.success(
              mapOf("event" to "onIceCandidate", "candidate" to candidate.asFlutterResult()))
        }

        override fun onNegotiationNeeded() {
          eventSink?.success(mapOf("event" to "onNegotiationNeeded"))
        }
      }

  init {
    chan.setMethodCallHandler(this)
    eventChannel.setStreamHandler(this)
    peer.addEventObserver(eventObserver)
    ControllerRegistry.register(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "createOffer" -> {
        scope.launch {
          try {
            result.success(peer.createOffer().asFlutterResult())
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "createAnswer" -> {
        scope.launch {
          try {
            result.success(peer.createAnswer().asFlutterResult())
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "setLocalDescription" -> {
        val descriptionArg: Map<String, Any>? = call.argument("description")
        val description =
            if (descriptionArg == null) {
              null
            } else {
              SessionDescription.fromMap(descriptionArg)
            }
        scope.launch {
          try {
            peer.setLocalDescription(description)
            result.success(null)
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "setRemoteDescription" -> {
        val descriptionArg: Map<String, Any> = call.argument("description")!!
        scope.launch {
          try {
            peer.setRemoteDescription(SessionDescription.fromMap(descriptionArg))
            result.success(null)
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "addIceCandidate" -> {
        val candidate: Map<String, Any> = call.argument("candidate")!!
        scope.launch {
          try {
            peer.addIceCandidate(IceCandidate.fromMap(candidate))
            result.success(null)
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "addTransceiver" -> {
        try {
          val mediaType = MediaType.fromInt(call.argument("mediaType")!!)
          val transceiverInitArg: Map<String, Any>? = call.argument("init")
          val transceiver =
              if (transceiverInitArg == null) {
                peer.addTransceiver(mediaType, null)
              } else {
                peer.addTransceiver(mediaType, RtpTransceiverInit.fromMap(transceiverInitArg))
              }

          val transceiverController = RtpTransceiverController(messenger, transceiver)
          result.success(transceiverController.asFlutterResult())
        } catch (e: Exception) {
          resultUnhandledException(result, e)
        }
      }
      "getTransceivers" -> {
        result.success(
            peer.getTransceivers().map {
              RtpTransceiverController(messenger, it).asFlutterResult()
            })
      }
      "restartIce" -> {
        peer.restartIce()
        result.success(null)
      }
      "getStats" -> {
        scope.launch {
          try {
            val stats = peer.getStats()
            if (stats == null) {
              result.success(listOf<Map<String, Any>>())
            } else {
              result.success(stats.asFlutterResult())
            }
          } catch (e: Exception) {
            resultUnhandledException(result, e)
          }
        }
      }
      "dispose" -> {
        disposeInternal(false)
        result.success(null)
      }
    }
  }

  override fun onListen(obj: Any?, sink: EventChannel.EventSink?) {
    if (sink != null) {
      eventSink = AnyThreadSink(sink)
    }
  }

  override fun onCancel(obj: Any?) {
    peer.removeEventObserver(eventObserver)
    eventChannel.setStreamHandler(null)
    eventSink?.endOfStream()
    eventSink = null
  }

  /**
   * Converts this [PeerConnectionController] to the Flutter's method call result.
   *
   * @return [Map] generated from this controller which can be returned to the Flutter side.
   */
  fun asFlutterResult(): Map<String, Any> =
      mapOf<String, Any>("channelId" to channelId, "id" to peer.id)

  override fun dispose() {
    disposeInternal(true)
  }

  /**
   * Closes method and event channels of this [PeerConnectionController].
   *
   * Disposes underlying [PeerConnectionProxy].
   */
  private fun disposeInternal(cancel: Boolean) {
    ControllerRegistry.unregister(this)
    scope.cancel("disposed")
    chan.setMethodCallHandler(null)
    peer.dispose()
    if (cancel) {
      onCancel(null)
    }
  }
}

File: ios/Classes/controller/PeerConnectionController.swift
import Flutter

/// Controller of a `PeerConnection`.
class PeerConnectionController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of the `PeerConnection`'s proxy'.
  private var peer: PeerConnectionProxy

  /// ID of the channel created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Controller of the `eventChannel` management.
  private var eventController: EventController

  /// Event channel for communicating with Flutter side.
  private var eventChannel: FlutterEventChannel

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Indicator whether this controller is disposed.
  private var isDisposed: Bool = false

  /// Initializes a new `PeerConnectionController` for the provided
  /// `PeerConnectionProxy`.
  init(messenger: FlutterBinaryMessenger, peer: PeerConnectionProxy) {
    let channelName = ChannelNameGenerator.name(
      name: "PeerConnection",
      id: self.channelId
    )
    self.eventController = EventController()
    self.messenger = messenger
    self.peer = peer
    self.peer.addEventObserver(
      eventObserver: PeerEventController(
        messenger: self.messenger, eventController: self.eventController
      )
    )
    self.channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: messenger
    )
    self.eventChannel = FlutterEventChannel(
      name: ChannelNameGenerator.name(
        name: "PeerConnectionEvent",
        id: self.channelId
      ),
      binaryMessenger: messenger
    )
    self.channel.setMethodCallHandler { call, result in
      self.onMethodCall(call: call, result: result)
    }
    self.eventChannel.setStreamHandler(self.eventController)
  }

  /// Sends the provided response to the provided result.
  ///
  /// Checks whether `FlutterMethodChannel` is not disposed before sending data.
  /// If it's disposed, then does nothing.
  func sendResultFromTask(_ result: @escaping FlutterResult, _ response: Any?) {
    if !self.isDisposed {
      result(response)
    }
  }

  /// Handles all the supported Flutter method calls for the controlled
  /// `PeerConnectionProxy`.
  func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "createOffer":
      Task {
        do {
          let sdp = try await self.peer.createOffer()
          self.sendResultFromTask(result, sdp.asFlutterResult())
        } catch {
          self.sendResultFromTask(result, getFlutterError(error))
        }
      }
    case "createAnswer":
      Task {
        let sdp = try! await self.peer.createAnswer()
        self.sendResultFromTask(result, sdp.asFlutterResult())
      }
    case "setLocalDescription":
      let description = argsMap!["description"] as? [String: Any]
      let type = description!["type"] as? Int
      let sdp = description!["description"] as? String
      Task {
        do {
          var desc: SessionDescription?
          if sdp == nil {
            desc = nil
          } else {
            desc = SessionDescription(
              type: SessionDescriptionType(rawValue: type!)!, description: sdp!
            )
          }
          try await self.peer.setLocalDescription(description: desc)
          self.sendResultFromTask(result, nil)
        } catch {
          self.sendResultFromTask(result, getFlutterError(error))
        }
      }
    case "setRemoteDescription":
      let descriptionMap = argsMap!["description"] as? [String: Any]
      let type = descriptionMap!["type"] as? Int
      let sdp = descriptionMap!["description"] as? String
      Task {
        do {
          try await self.peer.setRemoteDescription(
            description: SessionDescription(
              type: SessionDescriptionType(rawValue: type!)!,
              description: sdp!
            )
          )
          self.sendResultFromTask(result, nil)
        } catch {
          self.sendResultFromTask(result, getFlutterError(error))
        }
      }
    case "addIceCandidate":
      let candidateMap = argsMap!["candidate"] as? [String: Any]
      let sdpMid = candidateMap!["sdpMid"] as? String
      let sdpMLineIndex = candidateMap!["sdpMLineIndex"] as? Int
      let candidate = candidateMap!["candidate"] as? String
      Task {
        do {
          try await self.peer.addIceCandidate(
            candidate: IceCandidate(
              sdpMid: sdpMid!, sdpMLineIndex: sdpMLineIndex!,
              candidate: candidate!
            )
          )
          self.sendResultFromTask(result, nil)
        } catch {
          self.sendResultFromTask(result, getFlutterError(error))
        }
      }
    case "addTransceiver":
      let mediaType = argsMap!["mediaType"] as? Int
      let initArgs = argsMap!["init"] as? [String: Any]
      let direction = initArgs!["direction"] as? Int
      let argEncodings = initArgs!["sendEncodings"] as? [[String: Any]]
      var sendEncodings: [Encoding] = []

      for e in argEncodings! {
        let rid = e["rid"] as? String
        let active = e["active"] as? Bool
        let maxBitrate = e["maxBitrate"] as? Int
        let maxFramerate = e["maxFramerate"] as? Double
        let scaleResolutionDownBy = e["scaleResolutionDownBy"] as? Double
        let scalabilityMode = e["scalabilityMode"] as? String

        sendEncodings.append(Encoding(
          rid: rid!,
          active: active!,
          maxBitrate: maxBitrate,
          maxFramerate: maxFramerate,
          scaleResolutionDownBy: scaleResolutionDownBy,
          scalabilityMode: scalabilityMode
        ))
      }

      let transceiverInit =
        TransceiverInit(
          direction: TransceiverDirection(rawValue: direction!)!,
          encodings: sendEncodings
        )

      do {
        let transceiver = try RtpTransceiverController(
          messenger: self.messenger,
          transceiver: self.peer.addTransceiver(
            mediaType: MediaType(rawValue: mediaType!)!,
            transceiverInit: transceiverInit
          )
        )
        result(transceiver.asFlutterResult())
      } catch {
        result(getFlutterError(error))
      }
    case "getTransceivers":
      result(
        self.peer.getTransceivers().map {
          RtpTransceiverController(messenger: self.messenger, transceiver: $0)
            .asFlutterResult()
        }
      )
    case "restartIce":
      self.peer.restartIce()
      result(nil)
    case "dispose":
      self.isDisposed = true
      self.channel.setMethodCallHandler(nil)
      self.peer.dispose()
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller into a Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    [
      "channelId": self.channelId,
      "id": self.peer.getId(),
    ]
  }
}
